File: src/core/safety/runnerPolicy/planCompiler.ts
import type { RunnerExecutionPolicy, ExecutionPlan, Enforcement } from "./types";

export function compileRunnerExecutionPlan(policy: RunnerExecutionPolicy): ExecutionPlan {
  const enforcement: Enforcement[] = [];

  enforcement.push({
    area: "filesystem",
    strength: policy.filesystem.protectedPaths.length > 0 ? "Enforced" : "Unsupported",
    critical: policy.filesystem.protectedPaths.length > 0,
    description: `${policy.filesystem.protectedPaths.length} protected paths, scope: ${policy.filesystem.writableScope}`,
  });

  enforcement.push({
    area: "network",
    strength: policy.network.defaultAction === "deny" ? "Enforced" : "Advisory",
    critical: policy.network.defaultAction === "deny",
    description: `default: ${policy.network.defaultAction}, allowed: ${policy.network.allowedDomains.length}`,
  });

  enforcement.push({
    area: "credentials",
    strength: policy.credentials.blockOnSuspectedLeak ? "Enforced" : "Advisory",
    critical: policy.credentials.scanForLeaks,
    description: policy.credentials.blockOnSuspectedLeak ? "block on leak" : "scan only",
  });

  enforcement.push({
    area: "publish",
    strength: policy.publish.blockPushToMain ? "Enforced" : "Advisory",
    critical: policy.publish.blockPushToMain,
    description: `block-push-to-main: ${policy.publish.blockPushToMain}`,
  });

  enforcement.push({
    area: "commands",
    strength: policy.commands.blocked.length > 0 ? "Enforced" : "Advisory",
    critical: policy.commands.blocked.some(r => /rm\s+-rf|sudo/.test(r)),
    description: `${policy.commands.blocked.length} blocked, ${policy.commands.warnOnly.length} warn-only`,
  });

  enforcement.push({
    area: "external_sandbox",
    strength: policy.external_sandbox.enabled ? "Enforced" : "Unsupported",
    critical: false,
    description: policy.external_sandbox.tool ?? "disabled",
  });

  if (policy.preset === "Strict") {
    const violations = enforcement.filter(e =>
      e.critical && (e.strength === "Partial" || e.strength === "Unsupported"),
    );
    if (violations.length > 0) {
      return {
        policy: policy.preset,
        enforcement,
        rejected: true,
        rejection_reason: `Strict mode rejects critical enforcement with strength ${violations[0].strength} for area ${violations[0].area}`,
      };
    }
  }

  return { policy: policy.preset, enforcement, rejected: false };
}

File: crates/apohara-safety/src/runner_policy/plan_compiler.rs
//! Compiles a `RunnerExecutionPolicy` to an `ExecutionPlan` — ports
//! `src/core/safety/runnerPolicy/planCompiler.ts`.

use super::types::*;
use regex::Regex;
use std::sync::OnceLock;

pub fn compile_runner_execution_plan(policy: &RunnerExecutionPolicy) -> ExecutionPlan {
    let mut enforcement: Vec<Enforcement> = Vec::new();

    enforcement.push(Enforcement {
        area: EnforcementArea::Filesystem,
        strength: if !policy.filesystem.protected_paths.is_empty() {
            EnforcementStrength::Enforced
        } else {
            EnforcementStrength::Unsupported
        },
        critical: !policy.filesystem.protected_paths.is_empty(),
        description: Some(format!(
            "{} protected paths, scope: {}",
            policy.filesystem.protected_paths.len(),
            match policy.filesystem.writable_scope {
                WritableScope::Workspace => "workspace",
                WritableScope::Anywhere => "anywhere",
            }
        )),
    });

    enforcement.push(Enforcement {
        area: EnforcementArea::Network,
        strength: match policy.network.default_action {
            NetworkDefault::Deny => EnforcementStrength::Enforced,
            NetworkDefault::Allow => EnforcementStrength::Advisory,
        },
        critical: matches!(policy.network.default_action, NetworkDefault::Deny),
        description: Some(format!(
            "default: {}, allowed: {}",
            match policy.network.default_action {
                NetworkDefault::Allow => "allow",
                NetworkDefault::Deny => "deny",
            },
            policy.network.allowed_domains.len()
        )),
    });

    enforcement.push(Enforcement {
        area: EnforcementArea::Credentials,
        strength: if policy.credentials.block_on_suspected_leak {
            EnforcementStrength::Enforced
        } else {
            EnforcementStrength::Advisory
        },
        critical: policy.credentials.scan_for_leaks,
        description: Some(
            if policy.credentials.block_on_suspected_leak {
                "block on leak".to_string()
            } else {
                "scan only".to_string()
            },
        ),
    });

    enforcement.push(Enforcement {
        area: EnforcementArea::Publish,
        strength: if policy.publish.block_push_to_main {
            EnforcementStrength::Enforced
        } else {
            EnforcementStrength::Advisory
        },
        critical: policy.publish.block_push_to_main,
        description: Some(format!(
            "block-push-to-main: {}",
            policy.publish.block_push_to_main
        )),
    });

    enforcement.push(Enforcement {
        area: EnforcementArea::Commands,
        strength: if !policy.commands.blocked.is_empty() {
            EnforcementStrength::Enforced
        } else {
            EnforcementStrength::Advisory
        },
        critical: policy.commands.blocked.iter().any(|r| is_destructive_pattern(r)),
        description: Some(format!(
            "{} blocked, {} warn-only",
            policy.commands.blocked.len(),
            policy.commands.warn_only.len()
        )),
    });

    enforcement.push(Enforcement {
        area: EnforcementArea::ExternalSandbox,
        strength: if policy.external_sandbox.enabled {
            EnforcementStrength::Enforced
        } else {
            EnforcementStrength::Unsupported
        },
        critical: false,
        description: Some(match policy.external_sandbox.tool {
            Some(SandboxTool::Bwrap) => "bwrap".to_string(),
            Some(SandboxTool::Firejail) => "firejail".to_string(),
            None => "disabled".to_string(),
        }),
    });

    if matches!(policy.preset, PolicyPreset::Strict) {
        if let Some(violation) = enforcement.iter().find(|e| {
            e.critical
                && matches!(
                    e.strength,
                    EnforcementStrength::Partial | EnforcementStrength::Unsupported
                )
        }) {
            return ExecutionPlan {
                policy: policy.preset,
                rejection_reason: Some(format!(
                    "Strict mode rejects critical enforcement with strength {:?} for area {:?}",
                    violation.strength, violation.area
                )),
                enforcement,
                rejected: true,
            };
        }
    }

    ExecutionPlan {
        policy: policy.preset,
        enforcement,
        rejected: false,
        rejection_reason: None,
    }
}

fn is_destructive_pattern(r: &str) -> bool {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"rm\s+-rf|sudo").expect("static regex compiles"))
        .is_match(r)
}
